core/src/main/java/org/apache/iceberg/PartitionsTable.java

```diff
@@ -98,21 +105,46 @@ private DataTask task(TableScan scan) {
     return StaticDataTask.Row.of(partition.key, partition.recordCount, partition.fileCount);
   }
 
-  private static Iterable<Partition> partitions(Table table, Long snapshotId) {
+  private Iterable<Partition> partitions(StaticTableScan scan) {
+    CloseableIterable<FileScanTask> tasks = planTasks(scan);
+
     PartitionMap partitions = new PartitionMap(table.spec().partitionType());
-    TableScan scan = table.newScan();
+    for (FileScanTask task : tasks) {
+      partitions.get(task.file().partition()).update(task.file());
+    }
+    return partitions.all();
+  }
 
-    if (snapshotId != null) {
-      scan = scan.useSnapshot(snapshotId);
+  @VisibleForTesting
+  CloseableIterable<FileScanTask> planTasks(StaticTableScan scan) {
+    boolean caseSensitive = scan.isCaseSensitive();
+    boolean ignoreResiduals = scan.shouldIgnoreResiduals();
+    long snapshotId = scan.snapshot().snapshotId();
+
+    ManifestEvaluator eval = ManifestEvaluator.forPartitionFilter(
+        Projections.metadata(table.spec(), caseSensitive).project(scan.filter()),
```

Comment:
Is the purpose of the "metadata" projection just to rewrite predicates? For example, from `partition.x > 3` to `x > 3`?

I'm not sure that I would consider that a projection, but I need to think about it a bit more.

We can accomplish the same thing by creating an identity partition spec from the partition table's field names to the actual partition names. Or we could possibly just rewrite the predicates and remove `partition.`. Those options seem a bit simpler conceptually than using projection.

Reply:
Hi yes, it's basically this but it's a little more.  I should have explained a bit more my thinking in the PR.

Beyond transforming predicate "partition.x" => "x", there's also logic to handle predicates that don't have to do with the partitions, for example "record_count" > 0.  It seems the other Projections (InclusiveProjection/StrictProjection) were handling this logic (returning false for non-partition predicates, and all the and/or/not, etc),so I thought to re-use it.  It seems we have to 'project' those out, as  ManifestEvaluator Constructor tries bind predicates to the table's partitionSpec, and non-partition predicates will fail to be bound.

I can probably use a cloned partition spec as you said, but the problem would still remain, and I would have to write my own ExpressionVisitor that does the same logic?  Or let me know if I misunderstand.

Reply:
@szehon-ho, what I'm suggesting is that we can use the existing projections to do this rewrite work instead of creating a new projection. Using an existing projection with a different partition spec would be easier because you'd get the benefit you're talking about without adding a new form of projection that doesn't quite meet the definition of a projection.